Dynamic heights

soyuka opened the pull request:
This is now working :D.

I just need to add more tests on this particular point and check the reverse mode as I left some code commented.

soyuka commented on test/react.js:
```diff
@@ -54,7 +54,7 @@ describe('React', function () {
       const actual = renderedString
 
-      const expected = '<div style="width:100%;height:5px;overflow:auto;position:relative;" data-reactroot="" data-reactid="1" data-react-checksum="1386162411"><div style="opacity:0;position:absolute;width:1px;height:3px;" data-reactid="2"></div><div style="opacity:0;position:absolute;width:1px;height:3px;" data-reactid="3"></div><span style="position:absolute;top:0px;" class=" vrow" data-reactid ...
+      const expected = '<div style="width:100%;height:5px;overflow:auto;position:relative;" data-reactroot="" data-reactid="1" data-react-checksum="803780303"><div style="opacity:0;position:absolute;width:1px;height:3px;" data-reactid="2"></div><div style="opacity:0;position:absolute;width:1px;height:3px;" data-reactid="3"></div><span style="position:absolute;top:0px;" class=" vrow" data-reactid= ...
```
less rendered elements, the computation is now more precise

tbranyen replied:
Beautiful!

tbranyen commented on dist/hyperlist.js:
```diff
@@ -1,283 +0,0 @@
-(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.HyperList = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s( ...
```
I've been adding built files to each commit and using `.gitattributes` locally to suppress them in the diffs. It allows for very testing individual commits per SHA very easily. My recommendation is to create individual commits w/o the build and then the final commit (or flattened commits) contains the built representation of the work. Thoughts on this? I have some projects that omit and make disconnected commits that get tagged for release builds.

tbranyen replied:
I think by removing this file, you've removed the `dist` directory which is causing the CI to fail since `browserify` isn't going to create the folder for us.

soyuka replied:
Whoops, not my intention!

tbranyen commented on examples/2-million-rows.html:
```diff
@@ -132,7 +132,7 @@
     }
   };
 
-  var list = HyperList.default.create(tbody, config);
```
We duplicated this work, oh well it was easy enough to change. Thanks for this!

tbranyen replied:
Touche!

tbranyen commented on lib/index.js:
```diff
@@ -262,4 +306,56 @@ export default class HyperList {
     }
   }
+
+  // @TODO reverse?
```
Do we want to merge w/o this and just document it as a known caveat?

soyuka replied:
Nope I want this to work just like it was before, I'll test it today!

soyuka commented:
ping @tbranyen 

So, I fixed the reverse mode but the unit test there seems weird. I've added the row number to the chat example, if you could check it out, `0` is at the bottom and it does show every row in reverse. I think that the cause is that I'm still appending elements in their ascendant order but their positioning is reverse.

I still want to add some more unit tests on the dynamic heights but I think we're close to something good here!

Lmk if you have some comments/improvements!

tbranyen commented:
I'm going to test this out today w/ @sprjr (who has been requesting this feature) and let you know what we think. Makes sense about the reverse test, hopefully that's easy enough to fix (just look at style attrs).

soyuka commented:
Oh I can fix the test no problem there, just that I want to be sure you agree with the new behavior!

tbranyen commented:
I think it's fine, IMO the underlying insertion should be considered an implementation detail to the consumer and they should only care if the elements display in the correct position visually.

soyuka commented:
Okay fine, rebased and fixed the test by testing position instead of row!

tbranyen commented:
I have a patch waiting for once this gets merged to fix the remaining two examples: https://github.com/tbranyen/hyperlist/compare/dynamic-heights-examples-fix.patch I noticed that faker no longer ships w/ the unminified build so the examples broke.

tbranyen commented on dist/hyperlist.js:
```diff
@@ -210,57 +250,63 @@ var HyperList = function () {
+
+      var to = from + this._cachedItemsLen;
+      to = to > total ? total : to;
+      console.log(from, to);
```
Just a heads up you should rebuild the main package in this rebased final commit to remove this `console.log`.

soyuka commented:
Build + added a test feel free to merge!

sprjr commented:
Nice, this sounds really cool! I was chatting with @tbranyen earlier about the issue we ran into on another scroller - maybe it won't apply here but here was the run down:

- dynamic heights influenced the _actual_ height of the container
- you couldn't easily update the container height until after you rendered the items
- this caused scrolling to 'skip' some rows in larger lists as the rendered rows were expected to be a certain set of those, but because of the heights possibly being larger than anticipated the rendered would swap out the virtuals for the wrong offset

I'd love to test this out though and see if it'll cover that original use case. The other one we recently came across was making rows expandable but I don't think that's in the same space from an implementation standpoint (especially if a row height can be updated later anyway)

soyuka commented:
> dynamic heights influenced the actual height of the container

Not the case here, only the scroller height is changed accordingly.

> you couldn't easily update the container height until after you rendered the items

Here you can, you've to use the `refresh` method.

> but because of the heights possibly being larger than anticipated the rendered would swap out the virtuals for the wrong offset

Whoops, again not the case here this will just work.

> The other one we recently came across was making rows expandable but I don't think that's in the same space from an implementation standpoint (especially if a row height can be updated later anyway)

Would be doable, though I'd not recommend it, especially if you have a lot of elements. It may cause unwanted flickering. You may also be interested in [my comment here](https://github.com/tbranyen/hyperlist/issues/24#issuecomment-298367065).
